### PR TITLE
Reimplement tools/rw-heatmaps in go

### PR DESCRIPTION
**Pull request by ivanvc**
This PR translates Python's tools/rw-heatmaps to Go. It covers the use case in which the Python script takes a single input and plots a grid of heatmap (tripcolor) charts.

The Go implementation uses gonum.org/v1/plot, a plotting solution that could generate similar style charts while creating images. Unfortunately, it doesn't provide an equivalent to Python's tripcolor, so I'm doing a heatmap instead. While trying to keep the Go implementation similar to the original one, I used the closest color palette to the one that Python used (however, the color scale for Go's palette is inverted, so it needs to be inverted to have a similar color palette). 

I introduced a `go.mod` in the `tools/rw-heatmaps` directory. Because the `tools/` directory doesn't have a module defined. Therefore, any dependencies from this rewrite were getting into the top-level's `go.mod`. In the future, adding a module in `tools/` would be good if more tools are written in Go. Right now, the ones written in Go are not introducing any new dependencies, so the top level doesn't carry these dependencies.

### Python implementation charts

python_read
python_write

### Go implementation charts

go_read
go_write

### Used CSV

The CSV file used to generate the previous plots was this one: [result-202401270029.csv](https://github.com/etcd-io/etcd/files/14378480/result-202401270029.csv)

Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

File: tools/rw-heatmaps/README.md [hunk 1]
```diff
@@ -12,15 +12,11 @@ To get a mixed read/write performance evaluation result:
...
-To generate two images (read and write) based on the benchmark result csv file:
+To generate two images (read and write) based on the benchmark result CSV file:
 ```sh
 # to generate a pair of read & write images from one data csv file
-./plot_data.py ${FIRST_CSV_FILE} -t ${IMAGE_TITLE} -o ${OUTPUT_IMAGE_NAME}
-
-
-# to generate a pair of read & write images by comparing two data csv files
-./plot_data.py ${FIRST_CSV_FILE} ${SECOND_CSV_FILE} -t ${IMAGE_TITLE} -o ${OUTPUT_IMAGE_NAME}
+go run . ${CSV_FILE} -t ${IMAGE_TITLE} -o ${OUTPUT_IMAGE_NAME}
```
Comment (jmhbnz) [inline tools/rw-heatmaps/README.md:21]: Based on the help output it doesn't look like the short flags are in place yet?

```bash
  -h, --help                       help for rw-heatmaps
      --output-format string       output image file format (default "png")
      --output-image-file string   output image filename (required)
      --title string               plot graph title (required)

```

Comment (ivanvc) [reply tools/rw-heatmaps/README.md:21]: Right, I forgot to declare them :facepalm:

**Comment by ivanvc** [comment]
@jmhbnz, any feedback on the directory structure? Is it too much? Should I flatten `pkg/` subdirectories?

**Comment by jmhbnz** [comment]
> @jmhbnz, any feedback on the directory structure? Is it too much? Should I flatten `pkg/` subdirectories?

We don't have many files so feel free to flatten it for now. I'm pretty comfortable with it either way. Other reviewers may have stronger opinions on it, let's see.

File: tools/rw-heatmaps/main.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+import (
+	"os"
+
+	"go.etcd.io/etcd/tools/rw-heatmaps/v3/cmd"
+)
+
+func main() {
+	if err := cmd.NewRootCommand().Execute(); err != nil {
+		os.Exit(1)
```
[the remark this answers is not in the document]
Comment (ivanvc) [reply tools/rw-heatmaps/main.go:27]: Thanks for the suggestion. I applied it with an "error:" prefix to make it evident that there was an execution error.

**Comment by ivanvc** [comment]
> Thanks @ivanvc for the great work.
> 
> Please upload the CSV file(s) which you used to generate the heatmap diagrams above in this PR.

Hi @ahrtr, thank you. I uploaded it in the pull request description, but it is here too: https://github.com/etcd-io/etcd/files/14378480/result-202401270029.csv
> 
> Previous implementation has an performance comparison diagram as well, please refer to https://etcd.io/blog/2021/announcing-etcd-3.5/. But I do not see such diagram in this PR.
> 
I'll work on this, but I'm unfamiliar with how to generate this diagram, can you point me on how to do it? Seems like it wasn't covered in the original README.

**Comment by ahrtr** [comment]
> I'll work on this, but I'm unfamiliar with how to generate this diagram, can you point me on how to do it? Seems like it wasn't covered in the original README.

I am not familiar with that either. But technically speaking, I think it's should be same way as you generate the heatmap diagram above; you just need to replace the each value (displayed on the right y-axis) with (new value) / (old value).

**Comment by ahrtr** [comment]
> you just need to replace the each value (displayed on the right y-axis) with (new value) / (old value).

For example, assuming when connection is 32 and value size is 256, the throughput of the main branch is 1000 qps, while your PR is 1200, then use the value 1200/1000 = 1.2.

**Comment by ivanvc** [comment]
I finished porting the comparison charts to generate the blog post you mentioned @ahrtr. The following is the Go output of running the tool with the following two datasets:

* [result-202401270029.csv](https://github.com/etcd-io/etcd/files/14501708/result-202401270029.csv)
* [result-202402281701.csv](https://github.com/etcd-io/etcd/files/14501711/result-202402281701.csv)

Running

```sh
go run . -o comparison_go -t compare ../../result-202401270029.csv ../../result-202402281701.csv  -f png
```

Generates:

### Read
comparison_go_read

### Write
comparison_go_write

Passing the non-zero centered option:

```sh
go run . --zero-centered=false -o non_zero_centered_comparison_go -t compare ../../result-202401270029.csv ../../result-202402281701.csv -t png
```

Generates:

### Read
non_zero_centered_comparison_go_write

### Write
non_zero_centered_comparison_go_read

As comparison, this is the Python output:

### Read
python_comparison_write

### Write
python_comparison_read

By the way, I noticed a performance improvement in running this with Python and Go.

```sh
# Python
real    0m9.052s
user    0m8.520s
sys     0m2.628s
# Go
real    0m0.534s
user    0m0.672s
sys     0m0.151s
```

**Comment by ivanvc** [comment]
I squashed the commits, and I'm undrafting the PR.

**Comment by ivanvc** [comment]
@jmhbnz, a heads up that this is the first commit that fails the Go Vulnerability Checker because of the just-released [GO-2024-2611](https://pkg.go.dev/vuln/GO-2024-2611) (see https://github.com/etcd-io/etcd/actions/runs/8163213119/job/22315941148?pr=17428). This wasn't caught by Dependabot (as it was released hours ago). But can potentially block many PRs from now on. Would it make sense to create a new issue and have a PR soon to address it? I can work on this.

**Comment by jmhbnz** [comment]
> @jmhbnz, a heads up that this is the first commit that fails the Go Vulnerability Checker because of the just-released [GO-2024-2611](https://pkg.go.dev/vuln/GO-2024-2611) (see https://github.com/etcd-io/etcd/actions/runs/8163213119/job/22315941148?pr=17428). This wasn't caught by Dependabot (as it was released hours ago). But can potentially block many PRs from now on. Would it make sense to create a new issue and have a PR soon to address it? I can work on this.

If you have time please lend a hand with this weeks dependabot bumping and include an additional commit bumping that CVE impacted dependency also.

Feel free to ping me on k8s slack if you do want to tackle the weekly dependency bump and have any questions.  We have instructions and the rotation worksheet link here https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/dependency_management.md

**Comment by ahrtr** [comment]
Please rebase this PR.

**Comment by ivanvc** [comment]
Done @ahrtr

File: tools/rw-heatmaps/pkg/dataset/dataset.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+	// Param is the parameter used for the run.
+	Param string
+}
+
+// DataRecord holds the data for a single heatmap chart.
+type DataRecord struct {
+	ConnSize  int64
+	ValueSize int64
+	Read      float64
+	Write     float64
```
Comment (ahrtr) [inline tools/rw-heatmaps/pkg/dataset/dataset.go:53]:
```suggestion
	AvgRead      float64
	AvgWrite     float64
```

**Comment by ahrtr** [comment]
@ivanvc Great work! The code looks super clean and easy to understand.
- I did not read through the code under `pkg/chart`; instead I just compared a couple of values in the the CSV files and the images being generated, and they matches.
- I suggest to integrate the line charts in #15060 in the next step. They can share the same data structure defined in `pkg/dataset` in this PR. Please feel free to do it in a followup PR if you are interested. But we may want to rename the `rw-heatmaps` to something like `rw-benchmark`, in which there will be two visualisations: heatmap and line charts. All these can be in followup PRs.

File: tools/rw-heatmaps/README.md [hunk 4]
```diff
@@ -12,15 +12,35 @@ To get a mixed read/write performance evaluation result:
...
 ```sh
 # to generate a pair of read & write images from one data csv file
-./plot_data.py ${FIRST_CSV_FILE} -t ${IMAGE_TITLE} -o ${OUTPUT_IMAGE_NAME}
+go run . ${CSV_FILE} -t ${IMAGE_TITLE} -o ${OUTPUT_IMAGE_NAME}
+```
+
+To generate two images (read and write) showing the performance difference from two result CSV files:
+```sh
+# to generate a pair of read & write images from one data csv file
+go run . ${CSV_FILE1} ${CSV_FILE2} -t ${IMAGE_TITLE} -o ${OUTPUT_IMAGE_NAME}
```
Comment (ahrtr) [inline tools/rw-heatmaps/README.md:27]:
```suggestion
rw-heatmaps ${CSV_FILE1} ${CSV_FILE2} -t ${IMAGE_TITLE} -o ${OUTPUT_IMAGE_NAME}
```

File: tools/rw-heatmaps/README.md [hunk 5]
```diff
@@ -12,15 +12,35 @@ To get a mixed read/write performance evaluation result:
...
+```
+
+To see the available options use the `-h` option.
+
+```sh
+$ go run . -h
```
Comment (ahrtr) [inline tools/rw-heatmaps/README.md:33]:
```suggestion
$ rw-heatmaps -h
```

File: tools/rw-heatmaps/README.md (hunk 4 above)
Comment (ahrtr) [reply tools/rw-heatmaps/README.md:27]: Please also refer to https://github.com/etcd-io/etcd/blob/main/tools/benchmark/README.md to add installation and usage guide.

**Comment by ivanvc** [comment]
> @ivanvc Great work! The code looks super clean and easy to understand.
> 
>     * I did not read through the code under `pkg/chart`; instead I just compared a couple of values in the the CSV files and the images being generated, and they matches.
> 
>     * I suggest to integrate the line charts in [Reimplement the tool `rw-heatmaps` using golang and rename it to `rw-benchmark` #15060](https://github.com/etcd-io/etcd/pull/15060) in the next step. They can share the same data structure defined in `pkg/dataset` in this PR. Please feel free to do it in a followup PR if you are interested. But we may want to rename the `rw-heatmaps` to something like `rw-benchmark`, in which there will be two visualisations: heatmap and line charts. All these can be in followup PRs.

Thank you!

I can definitely work on the line charts in a follow-up PR.

File: tools/rw-heatmaps/README.md (hunk 4 above)
Comment (ivanvc) [reply tools/rw-heatmaps/README.md]: Done

File: tools/rw-heatmaps/README.md (hunk 5 above)
Comment (ivanvc) [reply tools/rw-heatmaps/README.md]: Done

File: tools/rw-heatmaps/pkg/dataset/dataset.go (hunk 3 above)
Comment (ivanvc) [reply tools/rw-heatmaps/pkg/dataset/dataset.go]: Done

File: tools/rw-heatmaps/README.md [hunk 6]
```diff
@@ -2,25 +2,69 @@
...
+Note: the result CSV file will be saved to current working directory. The working directory is where etcd database is saved. The working directory is designed for scenarios where a different mounted disk is preferred.
 
 ### Plot Graphs
-To generate two images (read and write) based on the benchmark result csv file:
+
+To generate two images (read and write) based on the benchmark result CSV file:
+
+```sh
+  # to generate a pair of read & write images from one data csv file
+  $ rw-heatmaps . ${CSV_FILE} -t ${IMAGE_TITLE} -o ${OUTPUT_IMAGE_NAME}
```
Comment (ahrtr) [inline tools/rw-heatmaps/README.md:44]:
```suggestion
  $ rw-heatmaps ${CSV_FILE} -t ${IMAGE_TITLE} -o ${OUTPUT_IMAGE_NAME}
```

File: tools/rw-heatmaps/README.md [hunk 7]
```diff
@@ -2,25 +2,69 @@
...
+```
+
+To generate two images (read and write) showing the performance difference from two result CSV files:
+
+```sh
+  # to generate a pair of read & write images from one data csv file
+  $ rw-heatmaps . ${CSV_FILE1} ${CSV_FILE2} -t ${IMAGE_TITLE} -o ${OUTPUT_IMAGE_NAME}
```
Comment (ahrtr) [inline tools/rw-heatmaps/README.md:51]:
```suggestion
  $ rw-heatmaps ${CSV_FILE1} ${CSV_FILE2} -t ${IMAGE_TITLE} -o ${OUTPUT_IMAGE_NAME}
```

File: tools/rw-heatmaps/README.md (hunk 6 above)
Comment (ivanvc) [reply tools/rw-heatmaps/README.md]: Oops, done.

**Review by ahrtr — APPROVED** [review]
lgtm

Thank you! Great work!

**Review by serathius — APPROVED** [review]

**Review by jmhbnz — APPROVED** [review]
This is looking great, thanks @ivanvc.

One suggestion that I'm happy to be addressed in a follow-up would be to add a `.gitignore` in the `rw-heatmaps` subdirectory to ensure users of the tool who run it from that directory (which may be fairly common) don't inadvertantly commit resulting files.

Here is what I see after building and running `rw-heatmaps`:

```bash
Untracked files:
  (use "git add <file>..." to include in what will be committed)
        result-202403111419.csv
        rw-heatmaps
        test.png_read.jpg
        test.png_write.jpg

```

Overall I think this is ready to merge and we can continue iterating with smaller pr's moving forward.

**Comment by ivanvc** [comment]
Let me know if you want me to add a `.gitignore` in this PR. That's fine with me.

**Comment by jmhbnz** [comment]
> Let me know if you want me to add a `.gitignore` in this PR. That's fine with me.

Let's leave this pr as is now to merge and come back with smaller follow-ups  🙏🏻

**Pull request merged**